chips/rp2040/src/pwm.rs

```diff
@@ -0,0 +1,1254 @@
+
+    // Configure the given channel using the given configuration
+    fn configure_channel(&self, channel_number: ChannelNumber, config: &PwmChannelConfiguration) {
+        self.set_ph_correct(channel_number, config.ph_correct);
+        self.set_invert_polarity(channel_number, config.a_inv, config.b_inv);
+        self.set_div_mode(channel_number, config.divmode);
+        self.set_divider_int_frac(channel_number, config.int, config.frac);
+        self.set_compare_value_a(channel_number, config.cc_a);
+        self.set_compare_value_b(channel_number, config.cc_b);
+        self.set_top(channel_number, config.top);
+        self.set_enabled(channel_number, config.en);
+    }
+
+    // Initialize the struct
+    fn init(&self) {
+        let default_config: PwmChannelConfiguration = PwmChannelConfiguration::default();
+        for channel_number in CHANNEL_NUMBERS {
+            self.configure_channel(channel_number, &default_config);
+            self.set_counter(channel_number, 0);
+            self.disable_interrupt(channel_number);
+        }
+        self.registers.intr.write(CH::CH.val(0));
+    }
+
+    // This method should be called when resolving dependencies for the
+    // default peripherals. See [crate::chip::Rp2040DefaultPeripherals::resolve_dependencies]
+    pub(crate) fn set_clocks(&self, clocks: &'a clocks::Clocks) {
+        self.clocks.set(clocks);
+    }
+
+    // Given a channel number and a channel pin, return a struct that allows controlling it
+    fn new_pwm_pin(&'a self, channel_number: ChannelNumber, channel_pin: ChannelPin) -> PwmPin<'a> {
+        PwmPin {
+            pwm_struct: self,
+            channel_number,
+            channel_pin,
+        }
+    }
+
+    // Map the given GPIO to a PWM channel and a PWM pin
+    fn gpio_to_pwm(&self, gpio: RPGpio) -> (ChannelNumber, ChannelPin) {
+        (ChannelNumber::from(gpio), ChannelPin::from(gpio))
+    }
+
+    /// Map the GPIO to a PwmPin struct
+    ///
+    /// The returned structure can be used to control the PWM pin.
+    ///
+    /// See [PwmPin]
+    pub fn gpio_to_pwm_pin(&'a self, gpio: RPGpio) -> PwmPin {
+        let (channel_number, channel_pin) = self.gpio_to_pwm(gpio);
+        self.new_pwm_pin(channel_number, channel_pin)
+    }
+
+    // Helper function to compute top, int and frac values
+    // selected_freq_hz ==> user's desired frequency
+    //
+    // Return value: Ok(top, int, frac) in case of no error, otherwise Err(())
+    fn compute_top_int_frac(&self, selected_freq_hz: usize) -> Result<(u16, u8, u8), ()> {
+        // If the selected frequency is high enough, then there is no need for a divider
+        // Note that unwrap can never fail.
+        let max_freq_hz = hil::pwm::Pwm::get_maximum_frequency_hz(self);
+        let threshold_freq_hz = max_freq_hz / hil::pwm::Pwm::get_maximum_duty_cycle(self);
+        if selected_freq_hz > threshold_freq_hz {
+            return Ok(((max_freq_hz / selected_freq_hz - 1) as u16, 1, 0));
```

Comment:
```suggestion
            return Ok(((max_freq_hz / (selected_freq_hz - 1)) as u16, 1, 0));
```
Is this what you meant here? if not, this subtraction can underflow for a number of different input frequencies depending on the value of `max_freq_hz`. I am not sure why you are subtracting 1 since I am not sure what the `top` value means. If you meant to divide, then subtract, change this to use `saturating_sub()` instead

Reply:
It's a division followed by a substraction. **top** is the maximum value of the counter. Once the counter reaches it, it eithers wraps to 0 or starts counting down (if running in double-edge mode).

Normally, no underflow should occur since one should choose the desired frequency based on the maximum value, i.e. `let freq = max_freq / 3`. I didn't add a constraint on this since I thought it is in the responsability of the caller to ensure the value is less than or equal to maximum value. But now I added a check to see if either the frequency is greater than **max_freq_hz** or if its value is 0.

Reply:
Makes sense, looks good to me